datafusion/core/src/datasource/physical_plan/file_scan_config.rs

```diff
@@ -194,6 +203,71 @@ impl FileScanConfig {
             .with_repartition_file_min_size(repartition_file_min_size)
             .repartition_file_groups(&file_groups)
     }
+
+    /// Attempts to do a bin-packing on files into file groups, such that any two files
+    /// in a file group are ordered and non-overlapping with respect to their statistics.
+    /// It will produce the smallest number of file groups possible.
```

Comment:
I understand the purpose of this is to split and group files.  Files in each group is sorted and not overlapped. If there are many groups, each of them must overlap with at least one other.

Is this the first step for following PRs to do something with these sorted file groups? Like files in each group will be read sequentially without going thru  SortPreservingMerge; and then data streams of different groups will go thru SortPreservingMerge?

Reply:
If I understand correctly what you're asking, I think it will already do this -- see for example the sqllogictest I added: https://github.com/apache/arrow-datafusion/blob/e9fad54257a13ab9630a0f038fa2b5a1b26aa006/datafusion/sqllogictest/test_files/parquet.slt#L193-L203

The `FileScanConfig` correctly detects that the files are ordered based on their statistics, and this information is propagated to the physical planner. So fortunately this works already. 

As for next steps, it actually just occurred to me that if there are 1000 files that all overlap, after this PR `ListingTable` will produce 1000 file groups which is really bad. So I think I should add a check that only uses the sorted file groups if they produce less than `target_partitions` file groups. I will add that in this PR. 

The other thing we should maybe do is modify `sort_file_groups` to _increase_ the number of file groups if it's lower than `target_partitions`, but I also thought maybe we should let DataFusion's `EnforceDistribution` physical optimizer take care of this, so it might not be necessary. 

Other than that I'm not sure immediately sure what next steps are, without gathering more data from real-world usage. Though, I would like to add some benchmarks to DataFusion for large plans (with thousands of files), as I think the performance of this code might become more apparent

Reply:
Thanks. I think another possible next step is to move the implementation of `ProgressiveEvalExec` from InfluxDB IOx here and then  someone can use it and the results of this PR to do https://github.com/apache/arrow-datafusion/issues/6672. 
What do you think @alamb ?

Reply:
> As for next steps, it actually just occurred to me that if there are 1000 files that all overlap, after this PR ListingTable will produce 1000 file groups which is really bad. So I think I should add a check that only uses the sorted file groups if they produce less than target_partitions file groups. I will add that in this PR.

I agree


> Thanks. I think another possible next step is to move the implementation of ProgressiveEvalExec from InfluxDB IOx here and then someone can use it and the results of this PR to do #6672.

I think it is a good idea @NGA-TRAN  -- @suremarc and anyone else following along `ProgressiveEvalExec` is an operator we implemented in InfluxDB 3.0 to replace `SortPreservingMerge` in plans where we know the inputs don't overlap in sort-key space (what this PR calculates) and that avoids the merge as well as reads the files more sequentially

> Other than that I'm not sure immediately sure what next steps are, without gathering more data from real-world usage. Though, I would like to add some benchmarks to DataFusion for large plans (with thousands of files), as I think the performance of this code might become more apparent

I agree a real world benchmark would be great. As the "running queries on sorted parquet files" is also quite important to InfluxDB 3.0 we can probably help you with that.

Do you have data in mind? If not, we could potentially make something using https://github.com/apache/arrow-datafusion/blob/main/test-utils/src/data_gen.rs#L251 (that sorts on the first few low cardinality columns, for example)

Reply:
@alamb I didn't have any data in mind, so any headstart we can get sounds great.

Regarding the `SortPreservingMerge`, I want to state again that `sort_file_groups` will produce (if possible) only one file group with all files sequentially ordered, so I think the plan will not have a sort-preserving-merge in the first place. I think maybe I just need to make this clearer in a separate sqllogictest. But crucially this means that #6672 would already be fixed by this PR.